design/ingressroute-v1-design.md

```diff
@@ -0,0 +1,320 @@
+
+**NOTE**: *Routing on a `header` value* still requires a path prefix to be specified. If pure header delegation is requested, then Contour will configure a wildcard path to match all paths resulting in a header match only across any path.
+
+#### Header Values
+
+IngressRoutev1 will support routing from various headers. It will implement commonly used headers such as user-agent & cookie since those will require a regex style implementation. By implementing them as speific types, we can apply those regex's automatically.
+
+- **header:** This is a generic header type and allows for matching custom headers
+- **cookie**: This is a built-in type and matches a `cookie` in the header while being implemented as a "contains" regex
+- **user-agent**: This is a built-in type and matches the `user-agent` in the header while being implemented as a "contains" regex
+
+### Path Match with Headers
+
+In the following example, you can define for path `/foo` and send traffic to different services for the same path but which contain different headers. This example shows how to have 2 routes match specific headers and a last route match a request without specific headers.
+
+```yaml
+spec:
+  routes:
+  - match:
+      path: /foo
+      - header: x-header
+        value: a
+    backends:
+    - http: backend-a
+      port: 9999
+  - match: 
+      path: /foo
+      - header: x-header
+        value: b
+    backends:
+    - http: backend-b
+      port: 9999
+  - match: 
+    path: /foo
+    backends:
+    - http: backend-default
+      port: 9999
+```
+
+#### Requests
+
+Following are sample requests and which backends will handle the request:
+
+- `GET -H "x-header: a" /foo` —> `backend-a`
+- `GET -H "x-header: b" /foo` —> `backend-b`
+- `GET /foo` —>  `backend-default`
+
+### Path Match with Cookie
+
+In the following example, you can route any request to path `/foo` containing a cookie containing `user=test` to the backend named `cookiebackend` over port `80`.
+
+```yaml
+spec:
+  routes:
+  - match:
+      path: /foo
+      - cookie: user=test
+      backends:
+      - http: cookiebackend
+        port: 80    
+```
+
+### Header Delegation
+
+Currently with IngressRoute, path prefixes of a request can be delegated to teams, users, namespaces, within a single Kubernetes cluster. Additionally, this design looks to add headers as a mechanism to pass authority off to teams, users or namespaces.
+
+The following example shows how requests to a specific path with a specific header can be delegated to a team in another namespace as well as how requests that don't match the headers specified previously will be handled in the current namespace.
+
+```yaml
+spec:
+  routes:
+  - match:
+      path: /foo
+      - header: x-header
+        value: a
+    delegate:
+      name: headera-ir
```

Comment:
What would the child ingressroute look like? Would it have the same match: stanza as the parent following the example we have in it beta1 where every path in the child ingressroute must share the same prefix as the parent?